clang/docs/checkedc/Bounds-Widening-for-Null-Terminated-Arrays.md

```diff
@@ -2,165 +2,290 @@
+  // If the Out set for the back edge of the loop is ∅, then the In set of the
+  // loop will always be ∅.
+  while (*p) {
+    // do something
+  }
+}
 ```
 
-## Implementation Details
-The main class that implements the analysis is
-[`BoundsAnalysis`](https://github.com/microsoft/checkedc-clang/blob/master/clang/lib/Sema/BoundsAnalysis.cpp)
-and the main function is `BoundsAnalysis::WidenBounds()`.
+So we maintain the set of all null-terminated array variables in the function
+and use it to initialize the `In` and `Out` sets for blocks.
+```
+Let the initial value of AllVars be ∅.
 
-`WidenBounds` will perform the bounds widening for the entire function. We can
-then call `BoundsAnalysis::GetWidenedBounds` to retrieve the widened bounds for
-the current basic block.
+∀ V in function F,
+  AllVars = AllVars ∪ {V}
+```
 
-The approach used for implementing the analysis is the iterative worklist
-algorithm in which we keep adding blocks to a worklist as long as we do not
-reach a fixed point i.e.: as long as the `Out` sets for the blocks keep changing.
+Thus, we initialize `In[B]` and `Out[B]` as follows:
+```
+∀ blocks B,
+  In[B] = Out[B] = {V:Top | V ∈ AllVars and Top is a special bounds expression}
+```
 
-### Algorithm
+Now, we also need to handle the case where there is an unconditional jump into
+a block. In this case, we cannot widen the bounds because we cannot provably
+infer that the element at `upper_bound(V)`
+is non-null on the unconditional edge. For example:
+```
+void f(_Nt_array_ptr<char> p, int x) {
+  if (condition) {
+    x = strlen(p) _Where p : bounds(p, p + x);
+  g(x); // p is not widened on all paths into this block.
+}
 ```
-1.  For the current function F:
-2.    For each basic block B in the reverse post-order for F:
-3.      Compute the Kill set for B
-4.      For each predecessor B' of B:
-5.        Compute the Gen set on edge B'->B
-6.      Add B to a queue called WorkList
 
-7.    For each basic block in WorkList:
-8.      Compute the In set for B
-9.      For each successor B' of B:
-10.       Store the current Out set Out[B][B'] as OldOut
-11.       Compute the new Out set on edge B->B'
-12.       Add B' to WorkList if Out[B][B'] != OldOut
+Thus, the `Out` set of the `Entry` block is propagated to all blocks that have
+at least one predecessor that does not widen the bounds of **any**
+null-terminated array. So in this case we want the intersection of `Out` blocks
+to be an empty set. To handle this case we initialize the `In` and `Out` sets
+of the `Entry` block to `∅`.
+```
+In[Entry] = Out[Entry] = ∅
 ```
 
-## Debugging the Analysis
-In order to debug the bounds widening anlaysis, you can use the clang flag
-`-fdump-widened-bounds`. This will dump the function name, the basic blocks
-sorted by block ID, and for each `_Nt_array_ptr` in the block the variable name
-and its widened bounds, if applicable.
+We also initialize the `In` and `Out` sets of all unreachable basic blocks to
+`∅`.
+
```

Comment:
You need a third pass that computes widened bounds for each statement, by walking through a block from the beginning to end, using In[S].   The implementation would provide a way of iterating through the statements in a basic block, computing the widened bounds after each statement.